packages/twenty-server/src/engine/core-modules/auth/auth.resolver.ts

```diff
@@ -111,6 +111,13 @@ export class AuthResolver {
       );
     }
     const user = await this.authService.challenge(challengeInput, workspace);
+
+    if (user.totpSecret)
+      throw new AuthException(
+        'OTP is required',
+        AuthExceptionCode.OTP_REQUIRED,
+      );
+
```

Comment:
After digging deeper into the code, I feel that `exceptions` are more reserved for "real errors". Things that shouldn't happen.

Maybe I'l switch to a different response for the `challenge` method. Something link `LoginToken | OTPRequired`, with
```tsx
type OTPRequired = { otpRequired: boolean }
```

Being inspired by `UserExists`

https://github.com/nicolasrouanne/twenty/blob/b24e43538dba1bedeb3ac2e2b1ff2197d3ee4c95/packages/twenty-server/src/engine/core-modules/auth/dto/user-exists.entity.ts#L5-L18

Reply:
Yes we lack a good framework. See: https://github.com/twentyhq/core-team-issues/issues/156 which should be tackle in the next few weeks